jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushImageStep.java

```diff
@@ -86,38 +86,32 @@
 
   @Override
   public BuildResult call() throws ExecutionException, InterruptedException {
-    ImmutableList.Builder<ListenableFuture<?>> dependenciesBuilder = ImmutableList.builder();
-    dependenciesBuilder.add(authenticatePushStep.getFuture());
-    for (AsyncStep<PushBlobStep> pushBlobStepStep : NonBlockingSteps.get(pushBaseImageLayersStep)) {
-      dependenciesBuilder.add(pushBlobStepStep.getFuture());
-    }
-    for (AsyncStep<PushBlobStep> pushBlobStepStep :
-        NonBlockingSteps.get(pushApplicationLayersStep)) {
-      dependenciesBuilder.add(pushBlobStepStep.getFuture());
-    }
-    dependenciesBuilder.add(NonBlockingSteps.get(pushContainerConfigurationStep).getFuture());
-    dependenciesBuilder.add(NonBlockingSteps.get(buildImageStep).getFuture());
-    return Futures.whenAllSucceed(dependenciesBuilder.build())
-        .call(this::afterPushSteps, listeningExecutorService)
+    return AsyncDependencies.using(listeningExecutorService)
+        .addStep(authenticatePushStep)
+        .addSteps(NonBlockingSteps.get(pushBaseImageLayersStep))
+        .addSteps(NonBlockingSteps.get(pushApplicationLayersStep))
+        .addStep(NonBlockingSteps.get(pushContainerConfigurationStep))
+        .addStep(NonBlockingSteps.get(buildImageStep))
+        .whenAllSucceed(this::afterPushSteps)
         .get()
         .get()
         .get();
   }
 
   private ListenableFuture<ListenableFuture<BuildResult>> afterPushSteps()
       throws ExecutionException {
-    List<ListenableFuture<?>> dependencies = new ArrayList<>();
-    for (AsyncStep<PushBlobStep> pushBlobStepStep : NonBlockingSteps.get(pushBaseImageLayersStep)) {
-      dependencies.add(NonBlockingSteps.get(pushBlobStepStep).getFuture());
+    AsyncDependencies dependencies = AsyncDependencies.using(listeningExecutorService);
+    for (AsyncStep<PushBlobStep> pushBaseImageLayerStep :
+        NonBlockingSteps.get(pushBaseImageLayersStep)) {
+      dependencies.addStep(NonBlockingSteps.get(pushBaseImageLayerStep));
     }
-    for (AsyncStep<PushBlobStep> pushBlobStepStep :
+    for (AsyncStep<PushBlobStep> pushApplicationLayerStep :
         NonBlockingSteps.get(pushApplicationLayersStep)) {
-      dependencies.add(NonBlockingSteps.get(pushBlobStepStep).getFuture());
+      dependencies.addStep(NonBlockingSteps.get(pushApplicationLayerStep));
```

Comment:
I wonder `NonBlockingSteps.get(pushApplicationLayersStep).stream().map(NonBlockingSteps::get).forEach(dependencies::addStep)` or something like may look better. Not that I insist using `stream()` though.

Reply:
The reason for not using `stream` is that `NonBlockingSteps.get` may throw an `ExecutionException` that can't be propagated directly from the lambda.